rpsl-tools/main.go

```diff
@@ -0,0 +1,186 @@
+	}
+	return nil
+}
+
+// getReader parses the filename and either returns a gzip reader or regular reader.
+func getReader(fn string) (io.Reader, error) {
+	fd, err := os.Open(fn)
+	if err != nil {
+		fmt.Printf("Failed to open file(%v): %v\n", fn, err)
+		return nil, err
+	}
+	if strings.HasSuffix(fn, ".gz") {
+		f, err := gzip.NewReader(fd)
+		if err != nil {
+			fmt.Printf("Failed to read gzip'd file(%v): %v\n", fn, err)
+			return nil, err
+		}
+		return f, nil
+	}
+	return fd, nil
+}
+
+// parseFile parses a single file at a time into records sent to the
+// records channel (rc). Files arrive as names on the input channel (ic).
+func parseFile(ic <-chan string, rc chan<- *rpsl.Record, ec chan<- bool) {
+	for fn := range ic {
+		var rdr *rpsl.Reader
+
+		reader, err := getReader(fn)
+		if err != nil {
+			glog.Infof("failed to read file(%v): %v\n", fn, err)
+			continue
+		}
+		rdr = rpsl.NewReader(reader)
+
+		// Read all leading comments and whitespace.
+		err = rdr.ConsumeComment()
+		if err != nil {
+			glog.Infof("Failed reading file(%v): %v\n", fn, err)
+			continue
+		}
+		err = rdr.ConsumeLeadingWS()
+		if err != nil {
+			glog.Infof("Failed reading file(%v): %v\n", fn, err)
+			continue
+		}
+
+		// The file must start with a letter, all IRR records start with a letter character.
+		r := rdr.Peek()
+		if !rpsl.IsLetter(r) {
+			glog.Infof("The first character read(%v) is not a letter, file unparsable.\n", string(r))
+			// Add 2 more chars so finding the problem is more possible.
+			r, _, _ := rdr.Read()
+			glog.Infof("Next char: %v\n", string(r))
+			r, _, _ = rdr.Read()
+			glog.Infof("Next char: %v\n", string(r))
+			continue
+		}
+
+		// Parse the file, sending results back up the channel (rc).
+		rpsl.Parse(rdr, rc)
+
+		// When done parsing, send a file-done message on the end channel (ec).
+		ec <- true
+	}
+}
+
+// Verify that there files requested exist, open each in a goroutine and feed
+// those to the rpsl-parser library, returning each record to a channel for
+// disposition in the final data structure to be loaded into a DB.
+func main() {
+	flag.Var(&rpslFiles, "rpslFiles", "Files to parse, irr/rpsl content(csv).")
+	flag.Parse()
+
+	// Flag checks.
+	if len(rpslFiles) == 0 {
+		flag.PrintDefaults()
+		return
+	}
+
+	// Two buffered channels, one for input and for the resulting records.
+	ic := make(chan string, len(rpslFiles))
+	rc := make(chan *rpsl.Record, 100)
+
+	// To signal that all files are done processing,
+	// use ec (end channel) to pass state of 'done with file X'
+	// to the Record processor.
+	ec := make(chan bool, len(rpslFiles))
+
+	// Start the parsing/worker thread
+	for i := 0; i < *threads; i++ {
+		go parseFile(ic, rc, ec)
+	}
+
+	// Push each file into the input (ic) channel.
+	for _, fn := range rpslFiles {
+		ic <- fn
+	}
+	close(ic)
+
+	// Track the number of files completed.
+	files := 0
+	// Read records from the channel,
+	// Note: I'm positive this is supposed to be simpler with
```

Comment:
use `TODO(morrowc)` notation in the comment?

Reply:
done